### ISSUE TITLE
postMessage API inconsistency

### ISSUE THREAD

**Issue by dsimmons** [slackapi/node-slack-sdk#291]
#### Description
Per looking at [this example](https://slackapi.github.io/node-slack-sdk/basic_usage#posting-a-message-with-web-api) in the documentation of the Web API, I expected my callback to have been called with an `Error` as the first argument if there's an issue.

### Expected result:
The `err` callback arg to have been populated with an `Error`.

### Actual result:
The `response` callback arg is called with an "error data structure".

Although I had an error handler in place, after a little bit of trial an error, I determined that the success handler was instead being called with the following: `{ ok: false, error: 'channel_not_found' }`.

For what it's worth, I triggered the above error scenario (trying to publish to a channel that doesn't exist) to double-check that my logging was working as expected.

Specifically, I'm wrapping `webClient.chat.postMessage` in a promise and, upon success, I'm acknowledging a message on a queue as processed. However, like mentioned above, I was erroneously acknowledging because I'd expected the error argument to have been provided.

I guess my question is this -- I feel like, from an API perspective, I should be handling ***one*** scenario or the other. In other words, if the error is always going to be contained within the `response` argument as an error data structure, then the callback only needs to have one param provided. Conversely, the library could invoke the callback with an `Error` if the response isn't successful. But, as of right now, I'm having to handle **both** possibilities:

```JavaScript
    // ... other code
    return new Promise((resolve, reject) => {
      webClient.chat.postMessage(channel, text, {
        as_user: true,
        attachments,
      }, (err, response) => {
        // NOTE: The following two lines specifically.
        const { error } = response;
        if (err || error) return reject(err || error);
        return resolve(response);
      });
    });
```

I guess I just more or less wanted confirm that this is as-intended from a design perspective, because it definitely caught me off-guard and sent me on a little 5-minute debugging detour! 😄

**Comment by dsimmons** [comment]
As a follow-up, I found the following mention in the [docs](https://api.slack.com/methods/chat.postMessage) for `chat.postMessage` specifically:

> This table lists the expected errors that this method could return. However, other errors can be returned in the case where the service is down or other unexpected factors affect processing. Callers should **always check the value of the ok params in the response**.

I admittedly must have missed that originally. Accordingly, per my above question, I'm assuming that it ***is*** intended from a design perspective. At the very least, I think it'd be worth highlighting in the `postMessage` code snippet I linked above -- at face value, it looks like the fact that `err` isn't populated, it can be assumed that the post was successful.

As far as changes outside of that, I'm still personally of the opinion that it could be made slightly more intuitive!

**Comment by aoberoi** [comment]
Hey @dsimmons, thanks for providing a detailed recount of your experience. This sort of thing really helps us understand where the rough edges are.

I agree that you shouldn't have to check both the truthiness of the error parameter and the truthiness of the `ok` property before deciding there was an error. I'll mark this as a bug so we can fix that. 

One thing I noticed is that you said you were wrapping the API call in a Promise, but you should know that when you don't pass in a callback, we already return a Promise. That might save you some time.

**Comment by dsimmons** [comment]
Great, I sincerely appreciate the quick and diplomatic response!

Also, thanks for making the note about the promise point -- ironically, I'd come across that in the documentation shortly after creating this issue 😄 

FWIW, just in terms of my personal opinion as an API consumer, major 👍  👍  👍  for exposing both callback and promise-based functions. The latter helps me eliminate unnecessary wrapper code, but I fully realize that there's a nontrivial sector of the JS community that still uses the former exclusively. So having it work as it does, it's definitely the best of both worlds! 💰

**Comment by DEGoodmanWilson** [comment]
It looks like there is an inconsistency in the way that promises vs callbacks are handled that is leading to this behavior? _I.e._, it looks like a bug? @aoberoi thoughts?
https://github.com/slackapi/node-slack-sdk/blob/master/lib/clients/client.js#L147

**Comment by DEGoodmanWilson** [comment]
@dsimmons Would you have a look at #296 and see if this helps resolve your issue?

**Comment by dsimmons** [comment]
FWIW I haven't tried running the code, but just quickly eyeballing the diff, that's exactly what I had in mind / had proposed above! Looks to unify the error handling with the callback approach.

I've since switched to using the promise interface (once I realized that was as thing), and like mentioned above, it doesn't appear to have said issue.

I do appreciate the swift response and you guys addressing this so quickly! 🎉 👍
